neuralcompression/functional/_lower_tail.py

```diff
@@ -3,49 +3,98 @@
+    @staticmethod
+    def forward(ctx, *args):
+        x, bound, gradient = args
+
+        bound = torch.tensor(bound, dtype=x.dtype)
+
+        ctx.save_for_backward(
+            x,
+            bound,
+            torch.tensor(gradient, dtype=torch.uint8),
+        )
+
+        return torch.max(x, bound)
+
+
+def lower_bound(
+    x: Tensor,
+    bound: Union[float, Tensor],
+    gradient: str = "identity_if_towards",
+) -> Tensor:
+    """``torch.maximum`` with a gradient for :math:`x < bound`.
+
+    This function is semantically equivalent to ``torch.maximum`` except the
+    behavior of the gradient with respect to :math:`x` for input values that
+    reach the :math:`bound` depends on the ``gradient`` option:
+
+        * ``"disconnected"``, the returned gradient is zero for values that
+          reach the :math:`bound`. This behavior is identical to the behavior
+          of ``torch.maximum``.
+
+        * ``"identity"``, the gradient is unconditionally replaced with the
+          identity function.
+
+        * ``"identity_if_towards"``, the gradient is replaced with the identity
+          function, but only if applying gradient descent would push the values
+          of inputs towards the :math:`bound`. For gradient values that push
+          away from the :math:`bound`, the returned gradient is still zero.
+
+    In the latter two cases, ``"identity"`` and ``"identity_if_towards"``, no
+    gradient is returned for :math:`bound`. Also, the implementation of
+    ``"identity_if_towards"`` assumes the shape of :math:`x` is the same as the
+    shape of the output (i.e. it won’t work reliably for all possible
+    broadcasting scenarios).
 
     Args:
-        distribution: an object representing a continuous-valued random
-            variable.
-        tail_mass: desired probability mass for the tails.
+        x: the input tensor.
+        bound: upper bound for :math:`x`.
+        gradient: The gradient to use. One of {``"disconnected"``,
+            ``"identity"``, ``"identity_if_towards"``}. Defaults to
+            ``"identity_if_towards"``.
 
     Returns:
-        the approximate lower tail quantiles for each scalar distribution.
+        the output tensor.
     """
-    try:
-        x = distribution.icdf(torch.tensor([tail_mass / 2]))
-    except (AttributeError, NotImplementedError):
-        try:
-            x = estimate_tails(
-                lambda _x: log_cdf(_x, distribution),
-                math.log(tail_mass / 2),
-                distribution.batch_shape,
-            )
-        except NotImplementedError:
-            message = """
-            `distribution` must implement either `cdf` or `icdf`
-            """
-
-            raise NotImplementedError(message)
-
-    return x.detach()
+
+    if gradient not in ("disconnected", "identity", "identity_if_towards"):
+        raise ValueError
+
+    return _LowerBound.apply(
+        x,
+        bound,
+        _LowerBoundGradient[gradient].value,
+    )
```

Comment:
Did you intend to move this from `_lower_bound.py`?

Reply:
Yep!

Reply:
aw, shit, no

Reply:
fixed